File: app/src/main/java/com/slapshotapps/dragonshockey/activities/admin/viewmodels/AdminGameViewModel.java
package com.slapshotapps.dragonshockey.activities.admin.viewmodels;

import android.content.Context;
import androidx.annotation.NonNull;
import com.slapshotapps.dragonshockey.R;
import com.slapshotapps.dragonshockey.Utils.DateFormaters;
import com.slapshotapps.dragonshockey.models.Game;
import com.slapshotapps.dragonshockey.models.GameResult;
import java.util.Date;

/**
 * Created on 10/16/16.
 */

public class AdminGameViewModel {
    private Game game;
    private Game originalGame;

    public AdminGameViewModel(@NonNull Game game) {
        this.game = game;

        try {
            originalGame = game.clone();
        }
        catch (CloneNotSupportedException e) {
            originalGame = new Game();
        }
    }

    public String getGameID(Context context) {
        return context.getString(R.string.edit_game_game_id, game.getGameID());
    }

    public String getOpponentName() {
        return game.getOpponent() == null ? "" : game.getOpponent();
    }

    public void setOpponentName(String name) {
        game.setOpponent(name);
    }

    public void setOpponentScore(String score) {
        if (game.getGameResult() == null) {
            createGameResult();
        }

        if (score == null || score.isEmpty()) {
            game.getGameResult().setOpponentScore(0);
        } else {
            game.getGameResult().setOpponentScore(Integer.valueOf(score));
        }
    }

    public String getOpponentScore() {
        if (game.getGameResult() != null) {
            return String.valueOf(game.getGameResult().getOpponentScore());
        } else {
            return "";
        }
    }

    public void setDragonsScore(String score) {
        if (game.getGameResult() == null) {
            createGameResult();
        }

        if (score == null || score.isEmpty()) {
            game.getGameResult().setDragonsScore(0);
        } else {
            game.getGameResult().setDragonsScore(Integer.valueOf(score));
        }
    }

    public String getDragonsScore() {
        if (game.getGameResult() != null) {
            return String.valueOf(game.getGameResult().getDragonsScore());
        } else {
            return "";
        }
    }

    public Date getGameDate() {
        return game.gameTimeToDate();
    }

    public void setGameDate(Date gameDate) {
        game.setGameTime(DateFormaters.convertDateToGameTime(gameDate));
    }

    public String getGameDateAsString() {
        return DateFormaters.getGameDate(game.gameTimeToDate());
    }

    public String getGameTimeAsString() {
        return DateFormaters.getGameTime(game.gameTimeToDate());
    }

    public boolean hasChanged() {
        return !originalGame.equals(game);
    }

    public Game getGame() {
        return game;
    }

    public boolean getOTL() {
        return game.getGameResult() != null && game.getGameResult().getOvertimeLoss();
    }

    public void setOTL(boolean isOTL) {
        if (game.getGameResult() == null) {
            createGameResult();
        }

        game.getGameResult().setOvertimeLoss(isOTL);
    }

    private void createGameResult() {
        game.setGameResult(new GameResult());
        game.getGameResult().setGameID(originalGame.getGameID());
    }
}

File: app/src/main/java/com/slapshotapps/dragonshockey/activities/admin/viewmodels/AdminGameViewModel.kt
package com.slapshotapps.dragonshockey.activities.admin.viewmodels

import android.content.Context
import com.slapshotapps.dragonshockey.R
import com.slapshotapps.dragonshockey.models.Game
import com.slapshotapps.dragonshockey.models.GameResult
import com.slapshotapps.dragonshockey.utils.DateFormatter.convertDateToGameTime
import com.slapshotapps.dragonshockey.utils.DateFormatter.getGameDate
import com.slapshotapps.dragonshockey.utils.DateFormatter.getGameTime
import java.util.Date

/**
 * Created on 10/16/16.
 */
class AdminGameViewModel(val game: Game) {
    private var originalGame: Game? = null

    init {
        originalGame = try {
            game.clone()
        } catch (e: CloneNotSupportedException) {
            Game()
        }
    }

    fun getGameID(context: Context): String {
        return context.getString(R.string.edit_game_game_id, game.gameID)
    }

    var opponentName: String?
        get() = if (game.opponent == null) "" else game.opponent
        set(name) {
            game.opponent = name
        }
    var opponentScore: String?
        get() = if (game.gameResult != null) {
            game.gameResult!!.opponentScore.toString()
        } else {
            ""
        }
        set(score) {
            if (game.gameResult == null) {
                createGameResult()
            }
            if (score == null || score.isEmpty()) {
                game.gameResult!!.opponentScore = 0
            } else {
                game.gameResult!!.opponentScore = Integer.valueOf(score)
            }
        }
    var dragonsScore: String?
        get() = if (game.gameResult != null) {
            game.gameResult!!.dragonsScore.toString()
        } else {
            ""
        }
        set(score) {
            if (game.gameResult == null) {
                createGameResult()
            }
            if (score == null || score.isEmpty()) {
                game.gameResult!!.dragonsScore = 0
            } else {
                game.gameResult!!.dragonsScore = Integer.valueOf(score)
            }
        }
    var gameDate: Date?
        get() = game.gameTimeToDate()
        set(gameDate) {
            game.gameTime = convertDateToGameTime(gameDate!!)
        }
    val gameDateAsString: String
        get() = getGameDate(game.gameTimeToDate()!!)
    val gameTimeAsString: String
        get() = getGameTime(game.gameTimeToDate()!!)

    fun hasChanged(): Boolean {
        return !originalGame!!.equals(game)
    }

    var oTL: Boolean
        get() = game.gameResult != null && game.gameResult!!.overtimeLoss
        set(isOTL) {
            if (game.gameResult == null) {
                createGameResult()
            }
            game.gameResult!!.overtimeLoss = isOTL
        }

    private fun createGameResult() {
        game.gameResult = GameResult()
        game.gameResult!!.gameID = originalGame!!.gameID
    }
}